text/0000-collection-view-redesign.md

```diff
@@ -0,0 +1,802 @@
+as much default implementation as possible, this means that ANY new method
+added to the collection library also involves unimplementing the method
+in the view library.   This can get particularly hairy as optimizations
+which get placed into "lower" collection classes must then be "undone" in the
+view hierarchy.   Views are constantly in danger of accidentally inheriting
+an operation which they cannot support.
+
+Furthermore, Each new node in the class tree requires extension of the memoized
+transformations that preceded it.  Here's an instance from `IndexedSeqViewLike`:
+
+```scala
+trait SeqViewLike[+A,
+                  +Coll,
+                  +This <: SeqView[A, Coll] with SeqViewLike[A, Coll, This]]
+  extends Seq[A] with SeqLike[A, This] with IterableView[A, Coll] with IterableViewLike[A, Coll, This]
+{ self =>
+  /** Explicit instantiation of the `Transformed` trait to reduce class file size in subclasses. */
+  private[collection] abstract class AbstractTransformed[+B] extends Seq[B] with super[IterableViewLike].Transformed[B] with Transformed[B]
+  trait Transformed[+B] extends SeqView[B, Coll] with super.Transformed[B] {
+    def length: Int
+    def apply(idx: Int): B
+    override def toString = viewToString
+  }  
+```
+
+This `Transformed` trait is one of the tricks used to ensure that views can stage
+operations and share implementations between each staged result.
+
+This results in bugs like [SI-4190](https://issues.scala-lang.org/browse/SI-4190)
+and [SI-4332](https://issues.scala-lang.org/browse/SI-4332)
+whereby the parent-behavior of a view is getting used through inheritance, when
+in fact we'd prefer a compile time error telling us to implement a missing method.
+
+```scala
+val x: ArrayBuffer[String] = ArrayBuffer("a", "b", "c")
+val y: IndexedSeq[String] = x.view map (_ + "0") // Class cast exception at runtime
+```
+
+
+## Abstract
+
+Here we'd like to propose a new view library with the following goals:
+
+* Views should solve the original goals for scala collections, sepcifically:
+  1. Calling `.view.force` on a collection returns something that is functionally
+      equivalent to the original collection, i.e. preserving as much of the
+      type as was known before.
+  2. Calling `.view`, executing a sequence of transformations, and then calling
+      `.force` should return a collection that is functionally equivalent to
+      just executing the sequence of transformations directly on the original
+      collection.
+  3. Calling `.view` and a sequence of transformations should not immediately
+     execute any transformation, but instead defer the operation, e.g. `map`,
+     `filter`, `slice`
+  4. Calling any operation which must be strictly evaluated will
+     result in immediate evaluation, e.g. `sum`, `fold`.
+* Views should NOT extend the collection API directly, rather they should
+  have their own interface, whereby implementations can be specifically tuned
+  for views.
```

Comment:
The issue here is actually that all of our "Abstractions" between things actually come with implementations riding along the way.  If we *had* a pure definition of an interface that you have to implement that was just a set of methods to call, it'd probably be ok.    

@retronym You may have seen the gitter discussion on this.  There is a solution to that problem, but the cost of it may be too high to consider.